components/cdc/src/rate_limiter.rs

```diff
@@ -0,0 +1,314 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::metrics::*;
+use crossbeam::channel::TrySendError;
+use std::cmp::min;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use tikv_util::mpsc::batch::Sender;
+
+#[derive(Debug)]
+pub enum RateLimiterError<E> {
+    SenderError(TrySendError<E>),
+    SinkClosedError(usize),
+    TryAgainError,
+}
+
+pub struct RateLimiter<E> {
```

Comment:
`RateLimiter` controls speed based on the length of the queue, where all pending messages are stored. If we used async-speed-limit, it would be be based on throughput rather than the queue length, making it very hard to respond to sudden decreases in network bandwidth or the downstream TiCDC's processing capacity. To address our current problems as well as futures problems (where TiCDC could suddenly stop reading TCP sockets because of insufficient memory), an adaptive mechanism is needed, and dynamically adjusting the throughput threshold used in async-speed-limit would involve complicated heuristics and a lot of testing.

To satisfy the need to (1) timely respond to changes in the queue's length and (2) release resources immediately if the queue is deemed too long to continue, we have to delve deep into the polling mechanism of the asynchronous channels and sinks, and handle congestion in the appropriate methods of the implementations of the sink.

A speed control based on throughput would be a nice addition, which can be easily added into the implementation of `RateLimiter`.

I will add more details comments on `RateLimiter` once I finish testing.